src/runtime/methodbinder.cs

```diff
@@ -350,192 +416,247 @@ internal Binding Bind(IntPtr inst, IntPtr args, IntPtr kw, MethodBase info, Meth
+                Debug.Assert(!isOperator);
+                if (pynargs == clrnargs - 1)
+                {
+                    fixedCnt = clrnargs - 1;
+                    matchType = MethodMatchType.Normal;
+                }
+                else if (pynargs < clrnargs - 1)
+                {
+                    // Included `foo(int x = 0, params object[] bar)`
+                    if ((pi[pynargs].Attributes & ParameterAttributes.HasDefault) == 0)
+                    {
+                        return false;
+                    }
+                    fixedCnt = pynargs;
+                    matchType = MethodMatchType.WithDefaultArgs;
+                }
+                else
+                {
+                    // This is a `foo(params object[] bar)` style method
+                    // Included `foo(int x = 0, params object[] bar)`
+                    fixedCnt = clrnargs - 1;
+                    matchType = MethodMatchType.WithParamArray;
+                }
             }
 
-            return clrtype;
-        }
+            for (int i = 0; i < fixedCnt; i++)
+            {
+                int fixedIdx = i + fixedStart;
+                ParameterInfo param = pi[fixedIdx];
+                Type clrType = param.ParameterType;
+                if (i >= pynargs)
+                {
+                    return false;
+                }
 
-        static bool MatchesArgumentCount(int argumentCount, ParameterInfo[] parameters,
-            out bool paramsArray,
-            out ArrayList defaultArgList)
-        {
-            defaultArgList = null;
-            var match = false;
-            paramsArray = false;
-
-            if (argumentCount == parameters.Length)
-            {
-                match = true;
-            } else if (argumentCount < parameters.Length)
-            {
-                match = true;
-                defaultArgList = new ArrayList();
-                for (var v = argumentCount; v < parameters.Length; v++) {
-                    if (parameters[v].DefaultValue == DBNull.Value) {
-                        match = false;
-                    } else {
-                        defaultArgList.Add(parameters[v].DefaultValue);
-                    }
+                IntPtr op = Runtime.PyTuple_GetItem(args, i);
+                if (!ExtractArgument(op, clrType, hasOverloads, ref margs[fixedIdx]))
+                {
+                    return false;
+                }
+
+                if (param.IsOut || clrType.IsByRef)
+                {
+                    outs++;
                 }
-            } else if (argumentCount > parameters.Length && parameters.Length > 0 &&
-                       Attribute.IsDefined(parameters[parameters.Length - 1], typeof(ParamArrayAttribute)))
-            {
-                // This is a `foo(params object[] bar)` style method
-                match = true;
-                paramsArray = true;
             }
 
-            return match;
+            switch (matchType)
```

Comment:
This method became very long. Can you cut it into smaller ones, so it would be easier to understand?